src/CodeAnalysis/CSharp/MetaAnalyzers/CSharpImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+            DiagnosticCategory.AnalyzerCorrectness,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return ImmutableArray.Create(DoNotIgnoreReturnValueDiagnosticRule);
+            }
+        }
+
+        private static readonly string s_solutionFullName = @"Microsoft.CodeAnalysis.Solution";
+        private static readonly string s_projectFullName = @"Microsoft.CodeAnalysis.Project";
+        private static readonly string s_documentFullName = @"Microsoft.CodeAnalysis.Document";
+        private static readonly string s_syntaxNodeFullName = @"Microsoft.CodeAnalysis.SyntaxNode";
+
+        // A list of known immutable object names
+        private static readonly List<string> s_immutableObjectNames = new List<string>()
+        {
+            s_solutionFullName,
+            s_projectFullName,
+            s_documentFullName,
+            s_syntaxNodeFullName,
+        };
+
+        private static readonly string s_Add = "Add";
+        private static readonly string s_Remove = "Remove";
+        private static readonly string s_Replace = "Replace";
+        private static readonly string s_With = "With";
+
+        private static readonly List<string> s_immutableMethodNames = new List<string>()
+        {
+            s_Add,
+            s_Remove,
+            s_Replace,
+            s_With,
+        };
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCodeBlockAction(AnalyzeCodeBlockForIgnoredReturnValues);
+        }
+
+        public void AnalyzeCodeBlockForIgnoredReturnValues(CodeBlockAnalysisContext context)
+        {
+            var model = context.SemanticModel;
+            var syntaxNode = context.CodeBlock;
+
+            //Find invocations that ignore the return value
+            //We're looking for invocations that are not within an equals value clause or a return statement.
+            var candidateInvocations = syntaxNode.DescendantNodes()
+                .OfType<InvocationExpressionSyntax>()
+                .Where(n => !(n.Ancestors().Any(m => m is EqualsValueClauseSyntax || m is ReturnStatementSyntax)));
+
+            foreach (var candidateInvocation in candidateInvocations)
+            {
+                //If we can't find the method symbol, quit
+                var methodSymbol = model.GetSymbolInfo(candidateInvocation).Symbol as IMethodSymbol;
+                if (methodSymbol == null)
+                    continue;
+
+                //If the method doesn't start with something like "With" or "Replace", quit
+                string methodName = methodSymbol.Name;
+                if (!s_immutableMethodNames.Any(n => methodName.StartsWith(n)))
```

Comment:
Actually, let me take that back. Do the simplest thing possible until you have measurements showing a performance problem. So I would leave the code as-is for now.

If you pass `/reportanalyzer` to csc.exe/vbc.exe they will spit out some information on the execution time of the analyzers. If nothing else this would give you some idea of how expensive this analyzer is relative to others in the package. But I won't require that kind of analysis in order to sign off on this change.